### ISSUE TITLE
argument validation errors should be thrown, not provided via callback

### ISSUE THREAD

**Issue by rjrodger** [senecajs/seneca#133]
they are calling code errors, after all

[Trello Card](https://trello.com/c/74VCQxkc/100-seneca-argument-validation-errors-should-be-thrown-not-provided-via-callback)

**Comment by baldmountain** [comment]
I have mixed feelings about this. Since you are exposing parambulator as the validation model for seneca.add/act, and parambulator doesn't throw errors, it provides errors through a callback, seneca should provide errors through a callback as well.

I should also say I'm not sure when to choose to return an Error as the first parameter of a callback vs. just throwing it.

**Comment by j** [comment]
@baldmountain I think that throwing Errors should be treated more as fatal and passing errors to the callback to be formatted for the response output.

I don't know why the response callback needs two arguments either and you can't just do response(new Error('some non-important error happened')) or response({ success: true }).

I'd be so stoked to be able to use Boom (https://github.com/hapijs/boom) style errors in the response callback like Hapi.js so that I can easily format errors and be consistent.  If anything, it'd be nice to be able to create my own Error instance and create consistent response outputs.

**Comment by geek** [comment]
Where does this change need to be made?  Our constructor and plugins both throw

**Comment by mcdonnelldean** [comment]
Needs discussion in the context of #398

**Comment by mcollina** [comment]
TL;DR I am 👎  on this.

We cannot make this throw unless we allow for a different behavior between local/in-memory microservices and distributed microservices. If they throw, I could crash another process by sending a wrongly composed message. A key principle of seneca is transport independence, even if there is _no_ transport, so I am against this change.

**Comment by baldmountain** [comment]
Having written and deployed lots of services using Seneca I also think that validation errors should **NOT** throw exceptions. Having said that, I usually skip over the parambulator based validation in Seneca and use Joi in actions. We needed some validations that the parambulator subset wouldn't do.

Now that I've said that, maybe a better feature would be to use Joi to validate parameters if the passed object is a Joi object and use the parambulator based validation if it is not.

But in either case, callback with an error if the parameters aren't valid.

**Comment by mcollina** [comment]
I agree we should switch to joi for validations. @mcdonnelldean ?

**Comment by mcdonnelldean** [comment]
There is already agreement on deprecating validation in favour of using your own (read Joi) I'll root out the issue on that. So I think thats just pending some of the bug work being complete.

On the errors, I'll ask for a bit more feedback on Gitter too. My personal preference is also to use the err callback.

My reasoning is thus, I don't want a bad validation to kill my service, I want my validation to enforce input. I don't see garbage input as an exception raising case rather a short circuit to tell the consumer they fecked up. 

Kindest Regards,

Dean

**Comment by kevin-madhu** [comment]
I found Joi joyful!

**Comment by mcdonnelldean** [comment]
Merged as this will be based on what happens with the error flow changes happening at the end of the year. This issue ties in with discussions on what constitutes an error vs an exception. All if this will be sounded out in #398

**Comment by rjrodger** [comment]
Agreed with discussion. Validation errors are provided via callback, not thrown.
If you look at both parambulator and joi, they work this way.

This issue is now more or less resolved with: https://github.com/senecajs/seneca/pull/420
See also: https://github.com/rjrodger/seneca-joi
